5_Stable_implementation/flip.cpp

```diff
@@ -0,0 +1,332 @@
+  std::ostringstream os;
+  bool first = true;
+  for (auto& tensor : tensors) {
+    if (tensor.defined()) {
+      if (!first) {
+        os << ", ";
+      }
+      os << tensor.sizes();
+      first = false;
+    }
+  }
+  return os.str();
+}
+
+static bool hasContiguousSubspace(torch::TensorList tl) {
+  // true if all the non-null tensors are adjacent
+  auto isDefined = [](const torch::Tensor & tensor){ return tensor.defined(); };
+  auto isNull = [](const torch::Tensor & tensor){ return !tensor.defined(); };
+  auto start = std::find_if(tl.begin(), tl.end(), isDefined);
+  auto stop = std::find_if(tl.rbegin(), tl.rend(), isDefined);
+  auto it = std::find_if(start, stop.base(), isNull);
+  return it == stop.base();
+}
+
+// Transposes the tensor and indices together so that all the non-null indices
+// index the first k dimensions of the tensor. Returns the transposed tensor
+// and the reordered indices. For example:
+// transposeToFront(tensor, {nullptr, a, nullptr, b})
+// returns
+// tensor.permute([1, 3, 0, 2]), {a, b, nullptr, nullptr}
+static std::tuple<torch::Tensor, std::vector<torch::Tensor>, std::vector<int64_t>>
+transposeToFront(torch::Tensor self, torch::TensorList indices) {
+  std::vector<int64_t> dims;
+  std::vector<int64_t> inv_dims(self.dim());
+  std::vector<torch::Tensor> transposedIndices;
+
+  dims.reserve(self.dim());
+  for (auto i = decltype(self.dim()){0}; i < self.dim(); i++) {
+    if (indices[i].defined()) {
+      inv_dims[i] = dims.size();
+      dims.push_back(i);
+      transposedIndices.emplace_back(indices[i]);
+    }
+  }
+  for (auto i = decltype(self.dim()){0}; i < self.dim(); i++) {
+    if (!indices[i].defined()) {
+      inv_dims[i] = dims.size();
+      dims.push_back(i);
+      transposedIndices.emplace_back();
+    }
+  }
+  auto permuted_input = self.permute(dims);
+
+  // std::reverse(dims.begin(), dims.end());
+  return std::make_tuple(permuted_input, std::move(transposedIndices), inv_dims);
+}
+
+// Replace indexed dimensions in src with stride 0 and the size of the result tensor.
+// The offset in these dimensions is computed by the kernel using the index tensor's
+// values and the stride of src. The new shape is not meaningful. It's used to make
+// the shape compatible with the result tensor.
+static torch::Tensor restride_src(const torch::Tensor& src, int64_t dims_before, int64_t dims_indexed,
+                                  torch::IntArrayRef replacement_shape) {
+  auto shape = torch::DimVector(src.sizes());
+  auto strides = torch::DimVector(src.strides());
+  int64_t end = dims_before + dims_indexed;
+  shape.erase(shape.begin() + dims_before, shape.begin() + end);
+  strides.erase(strides.begin() + dims_before, strides.begin() + end);
+  shape.insert(shape.begin() + dims_before, replacement_shape.begin(), replacement_shape.end());
+  strides.insert(strides.begin() + dims_before, replacement_shape.size(), 0);
+  return src.as_strided(shape, strides);
```

Comment:
I believe this can be simplified by just setting the strides of the dimensions to be indexed to be 0, instead of trying to take the indexer shapes into account. In the advanced indexing case, we don't know in advance the shape of the indexers so we need this, but in the code here we know it beforehand so we can leverage it right away (it will always be the same shape as the input)